### ISSUE TITLE
Eliminate code duplication in invalid assignment handling

### ISSUE THREAD

**Issue by stereotype441** [dart-lang/sdk#33913]
As far as I can tell, invalid assignments are handled by the front end in the following code paths:
- BodyBuilder.handleAssignmentExpression has code to detect assignments to non-generators
- BodyBuilder.handleUnaryPrefixAssignmentExpression has code to detect prefix increment of non-generators
- BodyBuilder.handleUnaryPostfixAssignmentExpression has code to detect postfix increment of non-generators
- DelayedAssignment.handleAssignment has code to detect assignment to constants
- DelayedPostfixIncrement.buildForEffect has code to detect postfix increment of constants in a void context
- DelayedPostfixIncrement.buildSimpleRead has code to detect postfix increment of constants in a non-void context
- ContextAwareGenerator.buildPrefixIncrement reports an error for prefix increment of assignments and postfix increments
- ErroneousExpressionGenerator.buildPrefixIncrement reports an error for prefix increment of unresolved names, improper uses of super or this, and improper accesses to instance fields from static contexts
- ThisAccessGenerator.buildPrefixIncrement reports an error for prefix increment of super or this
- ContextAwareGenerator.buildPostfixIncrement reports an error for postfix increment of assignments and postfix increments
- ErroneousExpressionGenerator.buildPostfixIncrement reports an error for postfix increment of unresolved names, improper uses of super or this, and improper accesses to instance fields from static contexts
- ThisAccessGenerator.buildPostfixIncrement reports an error for postfix increment of super or this
- ContextAwareGenerator.buildCompoundAssignment reports an error for compound assignment to assignments and postfix increments
- ErroneousExpressionGenerator.buildCompoundAssignment reports an error for compound assignment to improper uses of super or this, and improper accesses to instance fields from static contexts
- ThisAccessGenerator.buildCompoundAssignment reports an error for compound assignment to super or this
- KernelUnresolvedNameGenerator.buildCompoundAssignment reports an error for compound assignment to unresolved names
- ContextAwareGenerator.buildNullAwareAssignment reports an error for null-aware assignments to assignments and postfix increments
- ErroneousExpressionGenerator.buildNullAwareAssignment reports an error for null-aware assignments to unresolved names, improper uses of super or this, and improper accesses to instance fields from static contexts
- ThisAccessGenerator.buildNullAwareAssignment reports an error for null-aware assignments to super or this
- ContextAwareGenerator.buildAssignment reports an error for assignment to assignments and postfix increments
- ErroneousExpressionGenerator.buildAssignment reports an error for assignment to improper uses of super or this, and improper accesses to instance fields from static contexts
- ThisAccessGenerator.buildAssignment reports an error for assignment to super or this
- KernelUnresolvedNameGenerator.buildAssignment reports an error for assignment to unresolved names
- KernelLargeIntAccessGenerator._makeSimpleWrite reports an error for assignment to oversize ints
- KernelLargeIntAccessGenerator._makeWrite reports an error for compound assignment to oversize ints
- KernelVariableUseGenerator._makeWrite reports an error for assignment to const local variables
- KernelStaticAccessGenerator._makeWrite reports an error for assignment to const static variables

Rather than implement separate analyzer/front_end integration code paths for each of these cases (which would be a large amount of coding burden and would lead to a long tail of corner case bugs) I want to try to introduce some common code paths for handling each of these situations, so that less analyzer/FE integration needs to be written for these erroneous cases.  I believe code duplication in the front end should be reduced in the process.

Note that even though the task is to eliminate code duplication, I'm listing it as "type-bug" because not all of the above cases are handled by analyzer/FE integration yet, so unifying the code paths should fix analyzer/FE integration bugs.

**Comment by stefantsov** [comment]
I see your concerns, Paul, about possible future coding burden; however, to me, the cases listed above are an example of well-organized code, and I believe the code should not be refactored.  All of the items in the list, except for the first three, are methods of `ExpressionGenerator` sub-classes, designed to handle the process of generating expressions, and they are natural places where one would expect the assignment checks.  The remaining three items on the list are methods of `BodyBuilder` defined to handle assignments, and one should expect some error checking code there.  Also, I don't see how the logic of possible future coding burden could not be applied to the suggested refactoring itself.

/cc @kmillikin

**Comment by stereotype441** [comment]
I'll schedule a VC so we can discuss this.

**Comment by kmillikin** [comment]
Refactoring the front end like this is explicitly out of scope for the initial analyzer integration implementation.  Let's not do this now.

**Comment by devoncarew** [comment]
@kmillikin, I'm not sure what you intend to signal by closing this bug. Paul has already done the work here and landed the CL. We had to revert briefly to unblock Flutter, but fully intended to re-land it today.

We see refactoring this assignment handling as the only practical way to address all 27 of the specific cases that the bug mentions.  This issue is more about analyzer integration rather than solely about front end refactoring.

Let's plan to discuss at the scrum today.
